python/ray/rllib/dqn/dqn.py

```diff
@@ -262,67 +373,190 @@ def _init(self):
     def _update_worker_weights(self):
-        w = self.actor.get_weights()
-        weights = ray.put(self.actor.get_weights())
-        for w in self.workers:
-            w.set_weights.remote(weights)
+        if self.workers:
+            w = self.actor.get_weights()
+            weights = ray.put(self.actor.get_weights())
+            for w in self.workers:
+                w.set_weights.remote(weights)
 
     def _train(self):
+        if self.config["async_updates"]:
+            return self._train_async()
+        else:
+            return self._train_sync()
+
+    def _train_async(self):
+        apply_time = RunningStat(())
+        wait_time = RunningStat(())
+        gradient_lag = RunningStat(())
+        iter_init_timesteps = self.cur_timestep
+        num_gradients_applied = 0
+        gradient_list = [
+            worker.do_async_step.remote(
+                i, self.cur_timestep, self.actor.get_weights(),
+                num_gradients_applied)
+            for i, worker in enumerate(self.workers)]
+        steps = self.config["sample_batch_size"] * len(gradient_list)
+        self.cur_timestep += steps
+        self.steps_since_update += steps
+
+        while gradient_list:
+            dt = time.time()
+            gradient, info = ray.get(gradient_list[0])
+            gradient_list = gradient_list[1:]
+            wait_time.push(time.time() - dt)
+
+            if gradient is not None:
+                dt = time.time()
+                self.actor.apply_gradients(gradient)
+                apply_time.push(time.time() - dt)
+                gradient_lag.push(num_gradients_applied - info["gradient_id"])
+                num_gradients_applied += 1
+
+            if (self.cur_timestep - iter_init_timesteps <
+                    self.config["timesteps_per_iteration"]):
+                worker_id = info["id"]
+                gradient_list.append(
+                    self.workers[info["id"]].do_async_step.remote(
+                        worker_id, self.cur_timestep,
+                        self.actor.get_weights(), num_gradients_applied))
+                self.cur_timestep += self.config["sample_batch_size"]
+                self.steps_since_update += self.config["sample_batch_size"]
+
+            if (self.cur_timestep > self.config["learning_starts"] and
+                    self.steps_since_update >
+                    self.config["target_network_update_freq"]):
+                # Update target network periodically.
+                self.actor.dqn_graph.update_target(self.actor.sess)
+                self.steps_since_update -= (
+                    self.config["target_network_update_freq"])
+                self.num_target_updates += 1
+
+        mean_100ep_reward = 0.0
+        mean_100ep_length = 0.0
+        num_episodes = 0
+        buffer_size_sum = 0
+        stats = ray.get(
+            [w.stats.remote(self.cur_timestep) for w in self.workers])
+        for stat in stats:
+            mean_100ep_reward += stat[0]
```

Comment:
the indexing is not that extensible; would be nice to use dicts instead

Reply:
Added a TODO to clean this up.